### DISCUSSION TITLE
Is there a way to recall a cog slash command from a cog slash command?

### DISCUSSION THREAD

**Discussion by brotherelric** [discord-py-ui/discord-ui#101] [Q&A] [answered]
Example:
There is a setup cog slash command. When a user uses a command and it recognizes the user is new (based on json file), I want the command to "turn" into the setup command, running the code from the setup command instead.
```
if user in db:
      pass
else:
      [recall/redirect the user to setup slash command]
```
Thanks in advance.

**Comment by kvsxxx** [comment]
just call the callback of the setup command just like a normal method

For example
```py
class MyCog:
    ...
    @cogs.slash_cog(...)
    async def setup(self, ctx):
        # your command code
    @cogs.slash_cog(...)
    async def other_command(self, ctx):
        if user in db:
            pass
        else:
            # call the setup function
            await self.setup(ctx)
            return
```

This will run the setup command as if it was used as a slash command

**Reply by brotherelric** [reply]
As of what I'm testing, it can only work for one slash command right at a time, or am I wrong?
You would have to repeat this for every cog slash command in which I want the code from this to be recalled.
```python 
async def setup(self, ctx):
```

**Reply by kvsxxx** [reply]
what do you mean with repeat `async def setup(self, ctx):`? im kinda confused.

```py
async def setup(self, ctx):
```
is (in the example i send) your setup slash command you have.
As you said
> There is a setup cog slash command.

Sorry if I'm missunderstanding something but my english sucks

**Reply by brotherelric** [reply]
I may have not expressed myself correctly, my bad.
I was asking if there was any way to do it like this. Setup is a slash command separetely, which you can use. On the other hand, other commands may want to recall that slash command's code:
```python
@slash_cog(name="setup")
    async def setup(self, ctx):
        # setup command code

@slash_cog(name="profile")
    async def profile(self, ctx):
        if user in db:
            pass
        else:
            # call the setup function
            await self.setup(ctx)
            return

@slash_cog(name="command2")
    async def command2(self, ctx):
        if user in db:
            pass
        else:
            await self.setup(ctx)
            return
```

If I am understanding your response correctly, I wouldn't be able to do it like this, hence I would have to do it like this:
```python
@cogs.slash_cog(name="profile")
    async def setup(self, ctx):
        # setup command code
    async def profile(self, ctx):
        if user in db:
            pass
        else:
            # call the setup function
            await self.setup(ctx)
            return

@cogs.slash_cog(name="command2")
    async def setup(self, ctx):
        # setup command code
    async def command2(self, ctx):
        if user in db:
            pass
        else:
            await self.setup(ctx)
            return
```
If I am correct and that the only way is to do it like this, I don't think that would be any more efficient than just copy and pasting the code. I may be incorrect.

If I am not clear, let me know. Thanks!

**Reply by kvsxxx** [reply]
oh, I think I forgot the slash cog decorator in my example which maybe caused a bit of confusion. The first method of yours should work

(with the first method I mean)
```py
@slash_cog(name="setup")
async def setup(self, ctx):
    # setup command code

@slash_cog(name="profile")
async def profile(self, ctx):
    if user in db:
        pass
    else:
        # call the setup function
        await self.setup(ctx)
        return

@slash_cog(name="command2")
async def command2(self, ctx):
    if user in db:
        pass
    else:
        await self.setup(ctx)
        return

```

**Reply by brotherelric** [reply]
It isn't working for me unfortunately.
```
Ignoring exception in on_socket_response
Traceback (most recent call last):
  File "Python\Python310\lib\site-packages\discord\client.py", line 343, in _run_event
    await coro(*args, **kwargs)
  File "Python\Python310\lib\site-packages\discord_ui\client.py", line 244, in _on_slash_response
    await x.invoke(context, **options)
  File "Python\Python310\lib\site-packages\discord_ui\cogs.py", line 97, in invoke
    raise ex
  File "Python\Python310\lib\site-packages\discord_ui\cogs.py", line 92, in invoke
    await self.callback(self.cog, ctx, *args, **kwargs)
  File "\cogs\testcog.py", line 71, in test
    await self.setup(ctx)
  File "Python\Python310\lib\site-packages\discord_ui\cogs.py", line 74, in __call__
    return self.callback(*args, **kwds)
TypeError: testcog.setup() missing 1 required positional argument: 'ctx'
```

**Comment by kvsxxx** [comment, marked as answer]
okay, seems that there's an issue with directly calling cog commands with `()`, sorry for that, until it's fixed this should be a workaround

```py
...
await self.setup.invoke(ctx) # instead of await self.setup(ctx)
```

sorry for the trouble

**Reply by brotherelric** [reply]
Thank you! 😊
